cartography/intel/aws/kms.py

```diff
@@ -0,0 +1,333 @@
+            raise
+
+        described_key_list.append(response)
+
+    return described_key_list
+
+
+@timeit
+@aws_handle_regions
+def get_kms_key_details(boto3_session, kms_key_data, region):
+    """
+    Iterates over all KMS Keys.
+    """
+    client = boto3_session.client('kms', region_name=region)
+    for key in kms_key_data:
+        policy = get_policy(key, client)
+        aliases = get_aliases(key, client)
+        grants = get_grants(key, client)
+        yield key['KeyId'], policy, aliases, grants
+
+
+@timeit
+def get_policy(key, client):
+    """
+    Gets the KMS Key policy. Returns policy string or None if no policy
+    """
+    try:
+        policy = client.get_key_policy(KeyId=key["KeyId"], PolicyName='default')
+    except ClientError as e:
+        policy = None
+        logger.warning("Failed to retrieve Key Policy for key id - {}. Error - {}".format(key["KeyId"], e))
+        raise
+
+    return policy
+
+
+@timeit
+def get_aliases(key, client):
+    """
+    Gets the KMS Key Aliases.
+    """
+    aliases = []
+    paginator = client.get_paginator('list_aliases')
+    for page in paginator.paginate(KeyId=key['KeyId']):
+        aliases.extend(page['Aliases'])
+
+    return aliases
+
+
+@timeit
+def get_grants(key, client):
+    """
+    Gets the KMS Key Grants.
+    """
+    grants = []
+    paginator = client.get_paginator('list_grants')
+    for page in paginator.paginate(KeyId=key['KeyId']):
+        grants.extend(page['Grants'])
+
+    return grants
+
+
+@timeit
+def _load_kms_key_aliases(neo4j_session, aliases, update_tag):
+    """
+    Ingest KMS Aliases into neo4j.
+    """
+    ingest_aliases = """
+    MERGE (a:KMSAlias{id: {AliasArn}})
+    ON CREATE SET a.firstseen = timestamp(), a.targetkeyid = {TargetKeyId}
+    SET a.aliasname = {AliasName}, a.lastupdated = {UpdateTag}
+    WITH a MATCH (kmskey:KMSKey{id: {TargetKeyId}})
+    MERGE (a)-[r:KNOWN_AS]->(kmskey)
+    ON CREATE SET r.firstseen = timestamp()
+    SET r.lastupdated = {UpdateTag}
+    """
+
+    for alias in aliases:
+        neo4j_session.run(
+            ingest_aliases,
+            AliasArn=alias["AliasArn"],
+            AliasName=alias.get("AliasName"),
+            TargetKeyId=alias["TargetKeyId"],
+            UpdateTag=update_tag,
+        )
+
+
+@timeit
+def _load_kms_key_grants(neo4j_session, grants_list, update_tag):
+    """
+    Ingest KMS Key Grants into neo4j.
+    """
+    ingest_grants = """
+    MERGE (g:KMSGrant{id: {GrantId}})
+    ON CREATE SET g.firstseen = timestamp(), g.granteeprincipal = {GranteePrincipal}, g.creationdate = {CreationDate}
+    SET g.name = {GrantName}, g.lastupdated = {UpdateTag}
+    WITH g
+    MATCH (kmskey:KMSKey{id: {TargetKeyId}})
+    MERGE (g)-[r:APPLIED_ON]->(kmskey)
+    ON CREATE SET r.firstseen = timestamp()
+    SET r.lastupdated = {UpdateTag}
+    """
+
+    for grant in grants_list:
+        neo4j_session.run(
+            ingest_grants,
+            GrantId=grant["GrantId"],
```

Comment:
Is this based off an Arn?

Reply:
Each KMS Key Grant has an Id. but no ARN. That's why we are using the GrantId as the primary field.